src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -55,70 +58,175 @@ static void check(Coordination::Error code, const std::string & path)
-            if (dns_error)
-                throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
-            else
-                throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
+            /// availability_zones is empty on server startup or after config reloading
+            /// We will keep the az info when starting new sessions
+            availability_zones = args.availability_zones;
+            if (args.availability_zone_autodetect)
+                updateAvailabilityZones();
         }
+        chassert(availability_zones.size() == args.hosts.size());
+
+        /// Shuffle the hosts to distribute the load among ZooKeeper nodes.
+        std::vector<ShuffleHost> shuffled_hosts = shuffleHosts();
+        Coordination::ZooKeeper::Nodes nodes = hostsToNodes(log, shuffled_hosts);
 
         impl = std::make_unique<Coordination::ZooKeeper>(nodes, args, zk_log);
+        Int8 node_idx = impl->getConnectedNodeIdx();
 
         if (args.chroot.empty())
             LOG_TRACE(log, "Initialized, hosts: {}", fmt::join(args.hosts, ","));
         else
             LOG_TRACE(log, "Initialized, hosts: {}, chroot: {}", fmt::join(args.hosts, ","), args.chroot);
+
+
+        /// If the balancing strategy has an optimal node then it will be the first in the list
+        bool connected_to_suboptimal_node = node_idx != shuffled_hosts[0].original_index;
+        bool respect_az = !args.client_availability_zone.empty();
+        bool may_benefit_from_reconnecting = respect_az || args.get_priority_load_balancing.hasOptimalNode();
+        if (connected_to_suboptimal_node && may_benefit_from_reconnecting)
+        {
+            auto reconnect_timeout_sec = getSecondsUntilReconnect(args);
+            LOG_DEBUG(log, "Connected to a suboptimal ZooKeeper host ({}, index {})."
+                           " To preserve balance in ZooKeeper usage, this ZooKeeper session will expire in {} seconds",
+                      impl->getConnectedHostPort(), node_idx, reconnect_timeout_sec);
+
+            auto reconnect_task_holder = DB::Context::getGlobalContextInstance()->getSchedulePool().createTask("ZKReconnect", [this, optimal_host = shuffled_hosts[0]]()
+            {
+                try
+                {
+                    LOG_DEBUG(log, "Trying to connect to a more optimal node {}", optimal_host.host);
+                    Coordination::ZooKeeper::Nodes node;
+                    node.emplace_back(hostToNode(log, optimal_host));
+                    std::unique_ptr<Coordination::IKeeper> new_impl = std::make_unique<Coordination::ZooKeeper>(node, args, zk_log);
+                    Int8 new_node_idx = new_impl->getConnectedNodeIdx();
+
+                    /// Maybe the node was unavailable when getting AZs first time, update just in case
+                    if (args.availability_zone_autodetect)
```

Comment:
small optimization to only update this when the `avaiability[new_node_idx].empty()!` ?

Reply:
This optimization is too small, but makes sense